nemo/collections/nlp/models/machine_translation/megatron_adapter_nmt_model.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+from omegaconf.dictconfig import DictConfig
+from omegaconf.omegaconf import open_dict
+from pytorch_lightning.trainer.trainer import Trainer
+
+from nemo.collections.nlp.modules.common.megatron.parallel_adapters import ParallelLinearAdapterConfig
+from nemo.collections.nlp.models.machine_translation.megatron_nmt_model import MegatronNMTModel
+from nemo.collections.nlp.parts.nlp_overrides import NLPSaveRestoreConnector
+from nemo.core.classes.mixins import adapter_mixins
+from nemo.utils import logging
+
+__all__ = ["MegatronAdapterNMTModel"]
+
+
+class MegatronAdapterNMTModel(MegatronNMTModel):
+    """
+    Megatron Adapter NMT training
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer)
+
+        with open_dict(cfg):
+            cfg.megatron_amp_O2 = False
+            cfg.micro_batch_size = self.cfg.micro_batch_size
+            cfg.global_batch_size = self.cfg.global_batch_size
+            cfg.precision = trainer.precision
+
+        self.adapter_name_keys = ['adapter_1', 'adapter_2']
+        frozen_model = MegatronNMTModel.restore_from(
+            cfg.get('pretrained_model_path'),
+            trainer=trainer,
+            override_config_path=cfg,
+            save_restore_connector=NLPSaveRestoreConnector(),
+        )
+
+        # set the base model and enc_dec_model module
+        self.enc_dec_model = frozen_model.enc_dec_model
+        logging.info(f'Before adding adapters:\n{self.summarize()}')
+        self.freeze()
+        for _, module in self.enc_dec_model.enc_dec_model.named_modules():
+            if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                for adapter_key in self.adapter_name_keys:
+                    module.add_adapter(
+                        name=adapter_key,
+                        cfg=ParallelLinearAdapterConfig(
+                            in_features=cfg.hidden_size,
+                            dim=cfg.adapter_tuning.adapter_dim,
+                            norm_position='post',
```

Comment:
Per @titu1994 recommendation, it might be worth having this be configurable. pre-norm seems better in some large-model settings for me.